File: examples/state/inmemory-py/store.py
"""In-memory key/value state for rat-state-inmemory-py — mirrors the Go reference's
store.go MODEL (global monotonic revision, compare-and-set, append-only change log
for Watch) so the two impls stay in lockstep.
"""

import threading
from typing import Dict, List, Tuple


class Store:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._data: Dict[str, Tuple[bytes, int]] = {}  # key -> (value, revision)
        self._log: List[Tuple[str, bytes, int]] = []    # append-only (key, value, revision)
        self._rev = 0

    def get(self, key: str):
        with self._lock:
            if key not in self._data:
                return False, b"", 0
            value, rev = self._data[key]
            return True, value, rev

    def put(self, key: str, value: bytes, if_rev: int):
        """Compare-and-set. Returns (committed, revision): on commit, revision is the
        new rev; on conflict, revision is the current (conflicting) rev (no write).
        In-memory always knows the outcome, so UNKNOWN never arises."""
        with self._lock:
            cur_rev = self._data[key][1] if key in self._data else 0
            if if_rev > 0 and cur_rev != if_rev:
                return False, cur_rev
            self._rev += 1
            self._data[key] = (value, self._rev)
            self._log.append((key, value, self._rev))
            return True, self._rev

    def list(self, prefix: str) -> List[str]:
        with self._lock:
            return sorted(k for k in self._data if k.startswith(prefix))

    def watch_backlog(self, prefix: str, from_rev: int):
        """Change-log events with revision >= from_rev whose key has the prefix, in
        revision order. from_rev == 0 means 'from now' (no backlog) per the contract;
        the vectors use from_rev=1 to replay from the beginning."""
        with self._lock:
            if from_rev == 0:
                return []
            return [(k, v, r) for (k, v, r) in self._log if r >= from_rev and k.startswith(prefix)]

File: plugins/state/inmemory-go/store.go
// store.go — the in-memory key/value state with a global monotonic revision,
// compare-and-set, and an append-only change log for Watch.
package main

import (
	"sort"
	"strings"
	"sync"
)

type entry struct {
	value    []byte
	revision int64
}

// changeEvent is one append-only log record (all PUTs; this reference has no
// Delete RPC). The log preserves revision order, which is what Watch replays.
type changeEvent struct {
	key      string
	value    []byte
	revision int64
}

type store struct {
	mu   sync.Mutex
	data map[string]entry
	log  []changeEvent
	rev  int64 // global monotonic revision
}

func newStore() *store { return &store{data: map[string]entry{}} }

func (s *store) get(key string) (found bool, value []byte, revision int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.data[key]
	if !ok {
		return false, nil, 0
	}
	return true, e.value, e.revision
}

// put applies compare-and-set. ifRev > 0 requires the current revision to match;
// 0 is unconditional. Returns (committed, revision): on commit, revision is the new
// rev; on conflict, revision is the current (conflicting) rev and no write happens.
// An in-memory backend always knows the outcome, so UNKNOWN never arises.
func (s *store) put(key string, value []byte, ifRev int64) (committed bool, revision int64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	cur := s.data[key] // zero {nil,0} if absent
	if ifRev > 0 && cur.revision != ifRev {
		return false, cur.revision
	}
	s.rev++
	s.data[key] = entry{value: value, revision: s.rev}
	s.log = append(s.log, changeEvent{key: key, value: value, revision: s.rev})
	return true, s.rev
}

func (s *store) list(prefix string) []string {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]string, 0, len(s.data))
	for k := range s.data {
		if strings.HasPrefix(k, prefix) {
			out = append(out, k)
		}
	}
	sort.Strings(out) // deterministic for golden vectors
	return out
}

// watchBacklog returns change-log events with revision >= fromRev whose key has the
// prefix, in revision order. fromRev == 0 means "from now" (no backlog) per the
// contract; the conformance vectors use fromRev=1 to replay from the beginning.
func (s *store) watchBacklog(prefix string, fromRev int64) []changeEvent {
	s.mu.Lock()
	defer s.mu.Unlock()
	if fromRev == 0 {
		return nil
	}
	var out []changeEvent
	for _, e := range s.log {
		if e.revision >= fromRev && strings.HasPrefix(e.key, prefix) {
			out = append(out, e)
		}
	}
	return out
}
